### PR TITLE
BR-4083: (SI 76022) Sugar upgrader bloats to multiple GB of RAM durin…

### PR DESCRIPTION
**Pull request by alex-vlasov**
…g upgrade on large instances

Fix memory leak

### CODE REVIEW

**Comment by yellowandy** [comment]
Looks good to me

File: shadow.c [hunk 1]
```diff
@@ -555,7 +555,11 @@ static char *get_full_path(const char *filename TSRMLS_DC)
     	if(new_state.cwd) free(new_state.cwd);
         return NULL;
     }
-	return estrndup(new_state.cwd, new_state.cwd_length);
+    char *full_path = estrndup(new_state.cwd, new_state.cwd_length);
```
Comment (nathanhruby) [inline shadow.c]: I think indentation here is off by a bit due to preceding closing bracket having the wrong indentation.

**Comment by nathanhruby** [comment]
minor whitespace related comment, but functionally change looks ok

**Comment by alex-vlasov** [comment]
Fixed indentation and changed `free()` to `efree()`

**Comment by nathanhruby** [comment]
LGTM

File: shadow.c [hunk 2]
```diff
@@ -555,7 +555,11 @@ static char *get_full_path(const char *filename TSRMLS_DC)
     	if(new_state.cwd) free(new_state.cwd);
         return NULL;
     }
-	return estrndup(new_state.cwd, new_state.cwd_length);
+	char *full_path = estrndup(new_state.cwd, new_state.cwd_length);
+	if (new_state.cwd) {
+		efree(new_state.cwd);
```
Comment (yellowandy) [inline shadow.c]: I think 'free' would suffice as new_state.cwd is populated by strdup and not 'estrudup'.

Comment (alex-vlasov) [inline shadow.c]: I've changed `strdup` to `estrdup` instead. php-src encourages to use `e***` functions over standard, as they are used by the engine for proper memory management (e.g. memory limit calculation).

Comment (alex-vlasov) [inline shadow.c]: Disregarding my previous commit, tests were failing with `estrdup` approach, I changed it back to `free`.

**Comment by skymeyer** [comment]
Looks good, did we do a full review for any other potential leaks ? After QA sign off we probably should do a new release as well.

**Pull request merged**